File: qqbot-module-host/src/main/java/com/mieai/qqbot/module/host/SemanticVersion.java
package com.mieai.qqbot.module.host;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

record SemanticVersion(long major, long minor, long patch) implements Comparable<SemanticVersion> {
    private static final Pattern PATTERN = Pattern.compile(
            "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(?:[-+][0-9A-Za-z.-]+)?$");

    static SemanticVersion parse(String value) {
        Matcher matcher = PATTERN.matcher(value);
        if (!matcher.matches()) {
            throw new IllegalArgumentException("Invalid semantic version: " + value);
        }
        try {
            return new SemanticVersion(
                    Long.parseLong(matcher.group(1)),
                    Long.parseLong(matcher.group(2)),
                    Long.parseLong(matcher.group(3)));
        } catch (NumberFormatException exception) {
            throw new IllegalArgumentException("Semantic version component is too large: " + value, exception);
        }
    }

    @Override
    public int compareTo(SemanticVersion other) {
        int result = Long.compare(major, other.major);
        if (result == 0) result = Long.compare(minor, other.minor);
        if (result == 0) result = Long.compare(patch, other.patch);
        return result;
    }
}

File: qqbot-module-host/src/main/kotlin/com/mieai/qqbot/module/host/SemanticVersion.kt
package com.mieai.qqbot.module.host

private val SEMANTIC_VERSION_PATTERN = Regex("^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(?:[-+][0-9A-Za-z.-]+)?$")

internal data class SemanticVersion(val major: Long, val minor: Long, val patch: Long) : Comparable<SemanticVersion> {
    override fun compareTo(other: SemanticVersion): Int = compareValuesBy(this, other, SemanticVersion::major, SemanticVersion::minor, SemanticVersion::patch)

    companion object {
        fun parse(value: String): SemanticVersion {
            val match = SEMANTIC_VERSION_PATTERN.matchEntire(value)
                ?: throw IllegalArgumentException("Invalid semantic version: $value")
            return try {
                SemanticVersion(match.groupValues[1].toLong(), match.groupValues[2].toLong(), match.groupValues[3].toLong())
            } catch (error: NumberFormatException) {
                throw IllegalArgumentException("Semantic version component is too large: $value", error)
            }
        }
    }
}
